src/NUnitFramework/tests/Constraints/DefaultConstraintTests.cs

```diff
@@ -20,8 +20,10 @@ public DefaultConstraintTests()
         [Test] public void Success_Int() => AssertSuccess(default(int));
         [Test] public void Success_NullableInt() => AssertSuccess(default(int?));
         [Test] public void Success_Long() => AssertSuccess(default(long));
+#if !NETFRAMEWORK // the error message is different between class framework and newer .NET runtimes with optimized builds
```

Comment:
Curious. Do you recall how it's different? I could see stack traces changes, but I'm not sure what else could.
I'm a little concerned that this loses some test coverage on netfx runs. Is there a way we could keep the coverage of the the `DefaultStructWithOverriddenEquals` scenario here on optimized builds?

Reply:
Oops, I see now that you explained the difference. I'll have to take a closer look here. I'm still curious if we can avoid losing the netfx coverage here

Reply:
I was a bit surprised about this too but I would believe this is quite the corner case, I don't think it will much difference when deciphering a failed Assert in test log. As said, I would believe they were already getting the same message if they were using optimized code against the assert.

Reply:
I've pulled your branch and this test seems to pass for me locally on all platforms, including net462 if I remove the conditional compilation. Does it matter if the `ToString()` differs?

Reply:
@stevenaw here's the error if the assert is removed: https://github.com/nunit/nunit/actions/runs/4739738110/jobs/8414851760?pr=4350#step:8:423

```
Errors, Failures and Warnings

1) Failed : NUnit.Framework.Constraints.DefaultConstraintTests.Success_DefaultStructWithOverriddenEquals
  Expected: default
  But was:  NUnit.Framework.Constraints.DefaultConstraintTests+StructWithStrangeEquals
   at NUnit.Framework.Constraints.DefaultConstraintTests.AssertSuccess[T](T actual)
   at NUnit.Framework.Constraints.DefaultConstraintTests.Success_DefaultStructWithOverriddenEquals()

Run Settings
    Number of Test Workers: 2
    Work Directory: D:\a\nunit\nunit\src\NUnitFramework\tests\bin\Release\net462
    Internal Trace: Off
```

Reply:
@lahma Your comment says the error message is different, but these tests are supposed to pass, meaning error message is not relevant.
I checked out your branch and removed the `#if !NETFRAMEWORK` and I don't see any test failing.
I pushed your branch without that code to https://github.com/nunit/nunit/pull/4357 to see of it fails on the build server.

Reply:
So it does fail on the build server, but not on my machine?
```
dotnet test --configuration=Release --framework=net462 \
--filter="NUnit.Framework.Constraints.DefaultConstraintTests"
```
gives:
`Passed!  - Failed:     0, Passed:    16, Skipped:     0, Total:    16, Duration: 15 ms - nunit.framework.tests.dll (net462)`

Reply:
@manfred-brands yes this seems to be somehow machine dependent, should I remove the problematic altogether then?

Reply:
The existing implementation calls [RuntimeHelper.Equals](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.runtimehelpers.equals?view=netframework-4.6.2) which is the correct way as it ignores `Equals` overrides: `if both are the same value type and the values of their underlying memory are equal`.
So why does it fail on your machine and the build server, but only if the tests are optimized. As the previous un-optimized version passes.

Reply:
To make things even stranger, if I create a new .NET Console application with target `net462`, optimize = true and run in it in release mode: the code passes when I run it in console and in debugger I can see assert is fine.

```c#
using NUnit.Framework.Constraints;

var defaultConstraintTests = new DefaultConstraintTests();
defaultConstraintTests.Success_DefaultStructWithOverriddenEquals();
```

![image](https://user-images.githubusercontent.com/171892/233025731-ce8b6ca5-dbad-435e-8eb4-80f5abb6083a.png)

Reply:
I have upgraded my SDK to the same as the build server used, but it is still not failing for me. 
This is scary, code that works only on some machines.
@stevenaw Does this test fail for you?

Reply:
I removed the test case but kept the struct as it's referenced in equality test that isn't broken. I force pushed the version - not sure how you want your history but single commit probably cleaner as I guess you don't squash merge commits by default.

Reply:
That test case is covered by the ImmutableArray case and the weird Equatable implementation will confuse any reader. Please remove it.

Yes, single commit is preferred by me

Reply:
Done and force-pushed.